code-studio/src/js/videos.js

```diff
@@ -0,0 +1,312 @@
+  }
+  setupVideoFallback(options, width, height);
+  return video;
+};
+
+function onVideoEnded() {
+  $('.video-modal').trigger("ended");
+}
+
+function onYouTubeIframeAPIReady() {
+  // requires there be an iframe#video present on the page
+  var player = new YT.Player('video');
+  player.addEventListener('onStateChange', function (state) {
+    if (state.data === YT.PlayerState.ENDED) {
+      onVideoEnded();
+    }
+  });
+}
+
+function createVideo(options) {
+  return $('<iframe id="video"/>').addClass('video-player').attr({
+    src: options.src,
+    scrolling: 'no'
+  });
+}
+
+// Options include:
+//   src - the url to the video
+//   key - an uid.
+//   name - a string.
+//   redirect - the redirect page after the video is dismissed.
+//   onClose - actions to take after closing the video dialog, or immediately
+//             if the video isn't shown.
+window.showVideoDialog = function(options, forceShowVideo) {
+  if (forceShowVideo === undefined) {
+    forceShowVideo = false;
+  }
+
+  if (options.onClose === undefined) {
+    options.onClose = function () {};
+  }
+
+  if (dashboard.clientState.hasSeenVideo(options.key) && forceShowVideo === false) {
+    // Anything we were going to do when the video closed, we ought to do
+    // right now.
+    options.onClose();
+    if (options.redirect) {
+      window.location.href = options.redirect;
+    }
+    return;
+  }
+
+  upgradeInsecureOptions(options);
+  var widthRatio = 0.8;
+  var heightRatio = 0.8;
+  var aspectRatio = 16 / 9;
+
+  var body = $('<div/>');
+  var content = $('#notes-content').contents().clone();
+  content.find('.video-name').text(options.name);
+  body.append(content);
+
+  var video = createVideo(options);
+  body.append(video);
+
+  var notesDiv = $('<div id="notes-outer"><div id="notes"/></div>');
+  body.append(notesDiv);
+  getShowNotes(options.key, notesDiv.children('#notes'));
+
+  var dialog = new Dialog({ body: body, redirect : options.redirect });
+  var $div = $(dialog.div);
+  $div.addClass('video-modal');
+
+  $('.video-modal').on("remove", function () {
+    // Manually removing src to fix a continual playback bug in IE9
+    // https://github.com/code-dot-org/code-dot-org/pull/5277#issue-116253168
+    video.removeAttr('src');
+    options.onClose();
+    dashboard.clientState.recordVideoSeen(options.key);
+    // Raise an event that the dialog has been hidden, in case anything needs to
+    // play/respond to it.
+    var event = document.createEvent('Event');
+    event.initEvent('videoHidden', true, true);
+    document.dispatchEvent(event);
+  });
+
+  var tabHandler = function(event, ui) {
+    var tab = ui.tab || ui.newTab;  // Depends on event.
+    var videoElement = $('#video');
+    if (tab.find('a').attr('href') === "#video") {
+      // If it is the video page, restore the src
+      videoElement.attr('src', options.src);
+    } else {
+      video.removeAttr('src');
+      stopTrackingVideoJSProgress();
```

Comment:
I didn't change this, but dropping a note that this appears to be the source of the bug with the fallback player and the "Show Notes" tab.  For some reason, we remove the `src` attribute to hide the player?  When using the Flash fallback player, there is no such attribute, which explains why nothing happens to it.  I think `stopTrackingVideoJSProgress()` was originally supposed to help with this, but it doesn't do anything to pause/hide the video.